File: src/test/java/de/mwa/evopiaserver/EventsAcceptanceTest.java
package de.mwa.evopiaserver;


import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import de.mwa.evopiaserver.api.dto.EventDto;
import de.mwa.evopiaserver.api.dto.TagDto;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.json.AutoConfigureJson;
import org.springframework.boot.test.autoconfigure.json.AutoConfigureJsonTesters;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.util.TestPropertyValues;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.test.context.ContextConfiguration;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.util.List;

import static com.fasterxml.jackson.module.kotlin.ExtensionsKt.jacksonObjectMapper;
import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@Testcontainers
@AutoConfigureJson
@AutoConfigureJsonTesters
@ContextConfiguration(initializers = {EventsAcceptanceTest.Initializer.class})
public class EventsAcceptanceTest {

    @LocalServerPort
    private int port;

    @Autowired
    private TestRestTemplate restTemplate;

    @Autowired
    private RepositoryTestHelper repositoryTestHelper;

    @Container
    private static final PostgreSQLContainer postgreSQLContainer = new PostgreSQLContainer("postgres")
            .withDatabaseName("integration-tests-db")
            .withUsername("sa")
            .withPassword("sa");

    @AfterEach
    public void cleanup() {
        repositoryTestHelper.resetEventTagsTable();
        repositoryTestHelper.resetEventTable();
    }

    @Test
    public void should_get_events() {
        String events = getAllEventsString();

        assertThat(events).isEqualTo("[]");
    }

    @Test
    public void should_add_event_without_tag_and_get_it() throws JsonProcessingException {
        var body = "{" +
                "\"name\": \"nameIt\"," +
                "\"description\": \"desc\"," +
                "\"date\": \"2020\"," +
                "\"time\": \"18:00\"," +
                "\"place\": \"Berlin\"," +
                "\"imagePath\": \"img/path.jpg\"," +
                "\"tags\":[]" +
                "}";
        addEvent(body);

        List<EventDto> events = getAllEvents();
        assertThat(events).hasSize(1);
        var event = events.get(0);
        assertThat(event.getDate()).isEqualTo("2020");
        assertThat(event.getName()).isEqualTo("nameIt");
        assertThat(event.getDescription()).isEqualTo("desc");
        assertThat(event.getTime()).isEqualTo("18:00");
        assertThat(event.getPlace()).isEqualTo("Berlin");
        assertThat(event.getImagePath()).isEqualTo("img/path.jpg");
        assertThat(event.getTags()).isEmpty();
    }

    @Test
    public void should_add_event_with_tag_and_get_it() throws JsonProcessingException {
        var body = "{" +
                "\"name\": \"nameIt\"," +
                "\"description\": \"desc\"," +
                "\"date\": \"2020\"," +
                "\"time\": \"18:00\"," +
                "\"place\": \"Berlin\"," +
                "\"imagePath\": \"img/path.jpg\"," +
                "\"tags\":[ {\"name\":\"myTag\"}," +
                "{\"name\":\"mySecondTag\"}]" +
                "}";
        addEvent(body);

        List<EventDto> events = getAllEvents();
        assertThat(events).hasSize(1);
        var event = events.get(0);
        assertThat(event.getDate()).isEqualTo("2020");
        assertThat(event.getName()).isEqualTo("nameIt");
        assertThat(event.getDescription()).isEqualTo("desc");
        assertThat(event.getTime()).isEqualTo("18:00");
        assertThat(event.getPlace()).isEqualTo("Berlin");
        assertThat(event.getImagePath()).isEqualTo("img/path.jpg");
        assertThat(event.getTags()).containsOnly(new TagDto("myTag"), new TagDto("mySecondTag"));
    }

    @Test
    public void should_delete_event() throws JsonProcessingException {
        var body = "{" +
                "\"name\": \"nameIt\"," +
                "\"description\": \"desc\"," +
                "\"date\": \"2020\"," +
                "\"time\": \"18:00\"," +
                "\"place\": \"Berlin\"," +
                "\"imagePath\": \"img/path.jpg\"," +
                "\"tags\":[]" +
                "}";
        addEvent(body);

        List<EventDto> eventBefore = getAllEvents();
        assertThat(eventBefore).hasSize(1);
        var id = eventBefore.get(0).component1();

        deleteEvent(id);

        List<EventDto> events = getAllEvents();
        assertThat(events).isEmpty();

    }

    private String getAllEventsString() {
        var askingUrl = "http://localhost:" + port + "/v2/events";
        var askResponse = restTemplate.exchange
                (askingUrl, HttpMethod.GET, HttpEntityFactory.forTestUser(), String.class);

        assertThat(askResponse.getStatusCode())
                .as("Not a successful call: " + askResponse.getBody())
                .isEqualTo(HttpStatus.OK);
        return askResponse.getBody();
    }

    private List<EventDto> getAllEvents() throws JsonProcessingException {
        var askingUrl = "http://localhost:" + port + "/v2/events";
        var askResponse = restTemplate.exchange
                (askingUrl, HttpMethod.GET, HttpEntityFactory.forTestUser(), String.class);

        assertThat(askResponse.getStatusCode())
                .as("Not a successful call: " + askResponse.getBody())
                .isEqualTo(HttpStatus.OK);

        ObjectMapper mapper = jacksonObjectMapper();
        var jsonInput = askResponse.getBody();
        return mapper.readValue(jsonInput, new TypeReference<List<EventDto>>(){});
    }

    private void addEvent(String body) {
        var addingUrl = "http://localhost:" + port + "/v2/events/upsert";
        var addResponse = restTemplate.exchange
                (addingUrl, HttpMethod.POST, HttpEntityFactory.forTestUserWith(body), String.class);

        assertThat(addResponse.getStatusCode())
                .as("Not a successful call: " + addResponse.getBody())
                .isEqualTo(HttpStatus.OK);
    }

    private void deleteEvent(Integer id) {
        var addingUrl = "http://localhost:" + port + "/v2/events/" + id;
        var removeResponse = restTemplate.exchange
                (addingUrl, HttpMethod.DELETE, HttpEntityFactory.forTestUser(), String.class);

        assertThat(removeResponse.getStatusCode())
                .as("Not a successful call: " + removeResponse.getBody())
                .isEqualTo(HttpStatus.OK);
    }

    static class Initializer
            implements ApplicationContextInitializer<ConfigurableApplicationContext> {
        public void initialize(ConfigurableApplicationContext configurableApplicationContext) {
            TestPropertyValues.of(
                    "spring.datasource.url=" + postgreSQLContainer.getJdbcUrl(),
                    "spring.datasource.username=" + postgreSQLContainer.getUsername(),
                    "spring.datasource.password=" + postgreSQLContainer.getPassword()
            ).applyTo(configurableApplicationContext.getEnvironment());
        }
    }
}

File: src/test/kotlin/de/mwa/evopiaserver/EventsAcceptanceTest.kt
package de.mwa.evopiaserver


import de.mwa.evopiaserver.db.kotlin.de.mwa.evopiaserver.dto.EventDto
import de.mwa.evopiaserver.db.kotlin.de.mwa.evopiaserver.dto.TagDto
import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.server.testing.*
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test


@OptIn(ExperimentalCoroutinesApi::class)
class EventsAcceptanceTest : ServerTestSetup() {

    @Test
    fun should_get_events() = runTest {
        val events = getAllEvents()

        assertThat(events).isEmpty()
    }


    @Test
    fun should_add_event_without_tag_and_get_it() = testApplication {
        val body = "{" +
                "\"name\": \"nameIt\"," +
                "\"description\": \"desc\"," +
                "\"from\": \"2022-02-27T18:00:00.000\"," +
                "\"to\": \"2022-02-27T19:00:00.000\"," +
                "\"place\": \"Berlin\"," +
                "\"imagePath\": \"img/path.jpg\"," +
                "\"tags\":[]" +
                "}"

        addEvent(body)

        val events = getAllEvents()

        assertThat(events).hasSize(1)
        val event: EventDto = events[0]
        assertThat(event.from).isEqualTo("2022-02-27T18:00:00.000")
        assertThat(event.name).isEqualTo("nameIt")
        assertThat(event.description).isEqualTo("desc")
        assertThat(event.to).isEqualTo("2022-02-27T19:00:00.000")
        assertThat(event.place).isEqualTo("Berlin")
        assertThat(event.imagePath).isEqualTo("img/path.jpg")
        assertThat(event.tags).isEmpty()
    }

    @Test
    fun should_add_event_with_tag_and_get_it() = testApplication {
        repositoryTestHelper.addTag("ExistingTag1")
        repositoryTestHelper.addTag("ExistingTag2")

        val body = "{" +
                "\"id\":-1," +
                "\"name\":\"5 - Beachen\"," +
                "\"description\":\"Beachen\"," +
                "\"from\":\"2022-02-27T18:00:00.000\"," +
                "\"to\":\"2022-02-27T19:00:00.000\"," +
                "\"place\":\"Beachmitte\"," +
                "\"imagePath\":\"assets/icons/ball-volleyball.svg\"," +
                "\"tags\":[" +
                "{\"name\":\"ExistingTag1\"}," +
                "{\"name\":\"ExistingTag2\"}" +
                "]}"
        addEvent(body)

        val events = getAllEvents()

        assertThat(events).hasSize(1)
        val event: EventDto = events[0]
        assertThat(event.from).isEqualTo("2022-02-27T18:00:00.000")
        assertThat(event.name).isEqualTo("5 - Beachen")
        assertThat(event.description).isEqualTo("Beachen")
        assertThat(event.to).isEqualTo("2022-02-27T19:00:00.000")
        assertThat(event.place).isEqualTo("Beachmitte")
        assertThat(event.imagePath).isEqualTo("assets/icons/ball-volleyball.svg")
        assertThat(event.tags).containsOnly(TagDto("ExistingTag1"), TagDto("ExistingTag2"))
    }

    @Test
    fun should_delete_event() = testApplication {
        val body = "{" +
                "\"name\": \"nameIt\"," +
                "\"description\": \"desc\"," +
                "\"from\": \"2022-02-27T18:00:00.000\"," +
                "\"to\": \"2022-02-27T19:00:00.000\"," +
                "\"place\": \"Berlin\"," +
                "\"imagePath\": \"img/path.jpg\"," +
                "\"tags\":[]" +
                "}"
        addEvent(body)

        val eventBefore = getAllEvents()
        assertThat(eventBefore).hasSize(1)
        val id = eventBefore[0].component1()

        deleteEvent(id)

        val eventAfter = getAllEvents()
        assertThat(eventAfter).isEmpty()

    }

    @Test
    fun should_only_update_one_event() = runTest {
        val body = "{" +
                "\"name\": \"nameIt\"," +
                "\"description\": \"desc\"," +
                "\"from\": \"2022-02-27T18:00:00.000\"," +
                "\"to\": \"2022-02-27T19:00:00.000\"," +
                "\"place\": \"Berlin\"," +
                "\"imagePath\": \"img/path.jpg\"," +
                "\"tags\":[]" +
                "}"

        addEvent(body)
        addEvent(body)

        val events = getAllEvents()

        val newPlace = "Heeresbaeckerei"
        val changedPlaceBody = "{" +
                "\"id\": ${events.first().id}" +
                "\"name\": \"nameIt\"," +
                "\"description\": \"desc\"," +
                "\"from\": \"2022-02-27T18:00:00.000\"," +
                "\"to\": \"2022-02-27T19:00:00.000\"," +
                "\"place\": \"$newPlace\"," +
                "\"imagePath\": \"img/path.jpg\"," +
                "\"tags\":[]" +
                "}"

        addEvent(changedPlaceBody)

        val eventsAfterChange = getAllEvents()
        val places = eventsAfterChange.map { it.place }
        assertThat(places).containsExactlyInAnyOrder("Berlin", newPlace)
        assertThat(eventsAfterChange.size).isEqualTo(events.size)
    }

    private suspend fun getAllEvents(): List<EventDto> {
        val url = "http://localhost:8080/v3/events"

        val client = HttpClient(CIO)
        val askResponse = client.request(url)

        assertThat(askResponse.status)
            .isEqualTo(HttpStatusCode.OK)

        client.close()

        return Json.decodeFromString(askResponse.bodyAsText())
    }

    private suspend fun addEvent(_body: String): HttpResponse {
        val client = HttpClient(CIO)
        val url = Url("http://localhost:8080/v3/events/upsert")
        val response = client.request(url) {
            method = HttpMethod.Post
            contentType(ContentType("application", "json"))
            setBody(_body)
        }
        val y = response.bodyAsText()
        assertThat(response.status)
            .`as`(response.bodyAsText())
            .isEqualTo(HttpStatusCode.OK)

        assertThat(y).isNotEmpty
        return response
    }

    private suspend fun deleteEvent(id: Int): HttpResponse {
        val client = HttpClient(CIO)
        val url = "http://localhost:8080/v3/events/$id"
        val response = client.request(url) {
            method = HttpMethod.Delete
        }
        assertThat(response.status)
            .isEqualTo(HttpStatusCode.OK)

        return response
    }
}
